### PR TITLE
C, enh: consider name after "struct", "union", or "enum"

### PR DESCRIPTION
**Pull request by masatake**
Consider following source file:

	 1: struct point { float x; float y; };
	 2: struct point point;

When pressing M-. at `|' in "struct rectangle { struct poi|nt a, b
};", the tag for the line 1 should have higher priority than the tag
for the line 2.

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

File: citre-lang-c.el [hunk 1]
```diff
@@ -98,6 +98,14 @@
           (setq syntax (if (eq syntax 'member)
                            'callable-member
                          'function))))
+      (save-excursion
+        (goto-char (car bounds))
+        (when (progn (backward-sexp 1) (not (eq (point) (car bounds))))
```
Comment (AmaiKinono) [inline citre-lang-c.el:103]: Can we avoid using `backward-sexp`? I'm dealing with it a lot recently, some `forward-sexp-function`s will not take you to the previous symbol, but goes further to the opening delimiter of some bigger sexp (like in `web-mode`).

Maybe it works well in `c-mode`, but I consider it to be generally unreliable.

I think you could do one of:

- `looking-back` a regexp, like `(rx (group (seq symbol-start (* anything) symbol-end)) (* (or space "\n")))` I didn't test it though.
- Move backward blanks and then grab the symbol. You can use `puni--backward-blanks` in https://github.com/AmaiKinono/Tokimacs/blob/master/site-lisp/puni.el.

File: citre-lang-c.el [hunk 2]
```diff
@@ -158,6 +166,16 @@
          `(filter (or ,(citre-core-filter-kind "function")
                       ,(citre-core-filter-kind "macro"))
                   +)))
+       ('struct
```
Comment (AmaiKinono) [inline citre-lang-c.el:169]: Can we put an example in the comment here? Also, I don't understand the example you gave in the OP:

```
struct point point;
```

Do people really do this?

I don't know C well, so I'm gonna trust you. Please, write a clear example, or do nothing if the logic is obvious to C programmers.

Comment (masatake) [reply citre-lang-c.el:169]: > Do people really do this?

Yes.

```
[yamato@dev64]~/.citre% readtags -t upstream-linux.tags -Q '(and (eq? $kind "variable") (eq? $typeref (concat "struct:" $name))))' -l | wc -l
391
[yamato@dev64]~/.citre% readtags -t upstream-linux.tags -Q '(and (eq? $kind "member") (eq? $typeref (concat "struct:" $name))))' -l | wc -l 
2439
```

`struct point point` is so annoying and it appears so frequently. If I can remember correctly, you can write the same things with  "struct point" in go lang.

File: citre-lang-c.el (hunk 1 above)
Comment (masatake) [reply citre-lang-c.el]: Updated. I used looking-back.

File: citre-lang-c.el [hunk 3]
```diff
@@ -98,6 +98,14 @@
           (setq syntax (if (eq syntax 'member)
                            'callable-member
                          'function))))
+      (save-excursion
+        (goto-char (car bounds))
+        (when (looking-back (rx symbol-start
+                                (group (or "struct" "union" "enum"))
+                                (+ space)
+                                (* (or (syntax symbol) (syntax word))))
```
[the remark this answers is not in the document]
Comment (masatake) [reply citre-lang-c.el]: Updated.

**Comment by AmaiKinono** [comment]
Thanks!

**Pull request merged**